include/cuco/detail/static_set/static_set_ref.inl

```diff
@@ -0,0 +1,359 @@
+  using base_type  = static_set_ref<Key, Scope, KeyEqual, ProbingScheme, StorageRef>;
+  using ref_type   = static_set_ref<Key, Scope, KeyEqual, ProbingScheme, StorageRef, Operators...>;
+  using key_type   = typename base_type::key_type;
+  using value_type = typename base_type::value_type;
+
+  static constexpr auto cg_size      = base_type::cg_size;
+  static constexpr auto window_size  = base_type::window_size;
+  static constexpr auto thread_scope = base_type::thread_scope;
+
+ public:
+  /**
+   * @brief Inserts an element.
+   *
+   * @param value The element to insert
+   * @return True if the given element is successfully inserted
+   */
+  __device__ bool insert(value_type const& value) noexcept
+  {
+    ref_type& ref_    = static_cast<ref_type&>(*this);
+    auto probing_iter = ref_.probing_scheme_(value, ref_.storage_ref_.num_windows());
+
+    while (true) {
+      auto const window_slots = ref_.storage_ref_[*probing_iter];
+
+      // TODO: perf gain with #pragma unroll since num_windows is build time constant
+      for (auto& slot_content : window_slots) {
+        auto const eq_res = ref_.predicate_(slot_content, value);
+
+        // If the key is already in the container, return false
+        if (eq_res == detail::equal_result::EQUAL) { return false; }
+        if (eq_res == detail::equal_result::EMPTY) {
+          auto const intra_window_index = thrust::distance(window_slots.begin(), &slot_content);
+          switch (attempt_insert(
+            (ref_.storage_ref_.data() + *probing_iter)->data() + intra_window_index, value)) {
+            case insert_result::CONTINUE: continue;
+            case insert_result::SUCCESS: return true;
+            case insert_result::DUPLICATE: return false;
+          }
+        }
+      }
+      ++probing_iter;
+    }
+  }
+
+  /**
+   * @brief Inserts an element.
+   *
+   * @param group The Cooperative Group used to perform group insert
+   * @param value The element to insert
+   * @return True if the given element is successfully inserted
+   */
+  __device__ bool insert(cooperative_groups::thread_block_tile<cg_size> group,
+                         value_type const& value) noexcept
+  {
+    auto& ref_        = static_cast<ref_type&>(*this);
+    auto probing_iter = ref_.probing_scheme_(group, value, ref_.storage_ref_.num_windows());
+
+    while (true) {
+      auto const window_slots = ref_.storage_ref_[*probing_iter];
+
+      auto const [state, intra_window_index] = [&]() {
+        for (auto i = 0; i < window_size; ++i) {
+          switch (ref_.predicate_(window_slots[i], value)) {
+            case detail::equal_result::EMPTY: return cuco::pair{detail::equal_result::EMPTY, i};
+            case detail::equal_result::EQUAL: return cuco::pair{detail::equal_result::EQUAL, i};
+            default: continue;
+          }
+        }
+        return cuco::pair<detail::equal_result, int32_t>{detail::equal_result::UNEQUAL, -1};
```

Comment:
Where is this -1 coming from? Can we abstract it?

Reply:
It's a dummy that would never be used if the `equal_result` is `UNEQUAL`.